lib/block_view_lockups.go

```diff
@@ -0,0 +1,1784 @@
+	operationData := utxoOpsForTxn[operationIndex]
+	operationIndex--
+	if operationIndex < 0 {
+		return fmt.Errorf("_disconnectCoinLockupTransfer: Trying to revert OperationTypeCoinLockupTransfer " +
+			"but malformed utxoOpsForTxn")
+	}
+	if operationData.PrevSenderLockedBalanceEntry == nil || operationData.PrevSenderLockedBalanceEntry.isDeleted {
+		return fmt.Errorf("_disconnectCoinLockupTransfer: Trying to revert OperationTypeCoinLockupTransfer " +
+			"but found nil or deleted PrevSenderLockedBalanceEntry")
+	}
+	if operationData.PrevReceiverLockedBalanceEntry == nil || operationData.PrevReceiverLockedBalanceEntry.isDeleted {
+		return fmt.Errorf("_disconnectCoinLockupTransfer: Trying to revert OperationTypeCoinLockupTransfer " +
+			"but found nil or deleted PrevReceiverLockedBalanceEntry")
+	}
+
+	// Fetch the LockedBalanceEntries in the view.
+	senderLockedBalanceEntry := bav.GetLockedBalanceEntryForHODLerPKIDProfilePKIDExpirationTimestampNanoSecs(
+		operationData.PrevSenderLockedBalanceEntry.HODLerPKID,
+		operationData.PrevSenderLockedBalanceEntry.ProfilePKID,
+		operationData.PrevSenderLockedBalanceEntry.ExpirationTimestampNanoSecs)
+	receiverLockedBalanceEntry := bav.GetLockedBalanceEntryForHODLerPKIDProfilePKIDExpirationTimestampNanoSecs(
+		operationData.PrevReceiverLockedBalanceEntry.HODLerPKID,
+		operationData.PrevReceiverLockedBalanceEntry.ProfilePKID,
+		operationData.PrevReceiverLockedBalanceEntry.ExpirationTimestampNanoSecs)
+
+	// Ensure reverting the transaction won't cause the recipients balances to increase
+	// or cause the senders balances to decrease.
+	if operationData.PrevSenderLockedBalanceEntry.BalanceBaseUnits.Lt(&senderLockedBalanceEntry.BalanceBaseUnits) {
+		return fmt.Errorf("_disconnectCoinLockupTransfer: Reversion of coin lockup transfer would " +
+			"result in less coins for sender")
+	}
+	if operationData.PrevReceiverLockedBalanceEntry.BalanceBaseUnits.Gt(&receiverLockedBalanceEntry.BalanceBaseUnits) {
+		return fmt.Errorf("_disconnectCoinLockupTransfer: Reversion of coin lockup transfer would " +
+			"result in more coins for receiver")
+	}
+
+	// Set the balance entry mappings.
+	bav._setLockedBalanceEntry(operationData.PrevSenderLockedBalanceEntry)
+	bav._setLockedBalanceEntry(operationData.PrevReceiverLockedBalanceEntry)
+
+	// By here we only need to disconnect the basic transfer associated with the transaction.
+	basicTransferOps := utxoOpsForTxn[:operationIndex]
+	err := bav._disconnectBasicTransfer(currentTxn, txnHash, basicTransferOps, blockHeight)
+	if err != nil {
+		return errors.Wrapf(err, "_disconnectCoinLockupTransfer")
+	}
+
+	return nil
+}
+
+//
+// CoinUnlock Transaction Logic
+//
+
+func (bav *UtxoView) _connectCoinUnlock(
+	txn *MsgDeSoTxn, txHash *BlockHash,
+	blockHeight uint32, blockTimestamp int64,
+	verifySignatures bool) (_totalInput uint64, _totalOutput uint64, _utxoOps []*UtxoOperation, _err error) {
+	var utxoOpsForTxn []*UtxoOperation
```

Comment:
Meta comment after reviewing: Do we need to require a balance to be passed here? I can imagine it's just going to be a little annoying to implement where first you have to FETCH how much you want to unlock and THEN you can unlock it. It's also inviting floating point issues where maybe the UI is off by a single base unit and then this function fails.

If there's a good reason to require them to specify how much they want to unlock then I'm in favor of it. But it seemed better to either get rid of it or, perhaps better, to ignore it if they set the value to zero or something like that. The latter gives you best of both worlds. If you want to unlock a specific amount you can, otherwise it will just unlock everything that's available.